controllers/complianceeventsapi/server.go

```diff
@@ -471,6 +505,95 @@ func parseQueryArgs(queryArgs url.Values, isCSV bool) (*queryOptions, error) {
 		}
 	}
 
+	parsed, err := setAuthorizedClusters(ctx, db, parsed, userConfig)
+	if err != nil {
+		// ErrNoAccess needs queryOptions
+		return parsed, err
+	}
+
+	return parsed, nil
+}
+
+// setAuthorizedClusters verifies that if a cluster filter is provided,
+// the user has access to this filter. If no cluster filter is provided,
+// it sets the cluster filter to all managed clusters the user has access to.
+// If the user has no access, then ErrNoAccess is returned.
+func setAuthorizedClusters(ctx context.Context, db *sql.DB, parsed *queryOptions,
+	userConfig *rest.Config,
+) (*queryOptions, error) {
+	unAuthorizedClusters := []string{}
+
+	// Get all managedCluster rules
+	allRules, err := getManagedClusterRules(userConfig, nil)
+	if err != nil {
+		return parsed, err
+	}
+
+	if slices.Contains(allRules["*"], "get") || slices.Contains(allRules["*"], "*") {
+		return parsed, nil
+	}
+
+	tempClusterIDs := parsed.Filters["clusters.cluster_id"]
+	clusterIDs := make([]string, len(tempClusterIDs))
+
+	// Deep copy parsed.Filters["clusters.cluster_id"]
+	copy(clusterIDs, tempClusterIDs)
+
+	// Reset clusters.cluster_id
+	parsed.Filters["clusters.cluster_id"] = []string{}
+	// Convert id to name
+	for _, id := range clusterIDs {
+		clusterName, err := getClusterNameFromID(ctx, db, id)
+		if err != nil {
+			if errors.Is(err, sql.ErrNoRows) {
+				// Filter out invalid cluster IDs from the query
+				continue
+			}
+
+			log.Error(err, "Failed to get cluster name from cluster ID", "ID", id)
+
+			return parsed, err
+		}
+
+		if !getAccessByClusterName(allRules, clusterName) {
+			unAuthorizedClusters = append(unAuthorizedClusters, id)
+		} else {
+			parsed.Filters["clusters.cluster_id"] = append(parsed.Filters["clusters.cluster_id"], id)
+		}
+	}
+
+	parsedClusterNames := parsed.Filters["clusters.name"]
+	for _, clusterName := range parsedClusterNames {
+		if !getAccessByClusterName(allRules, clusterName) {
+			unAuthorizedClusters = append(unAuthorizedClusters, clusterName)
+		}
+	}
+
+	// There is no cluster.id or clusterName arg from the url query
+	// In other words, the user requests all they have access to.
+	if len(clusterIDs) == 0 && len(parsedClusterNames) == 0 {
+		starRules, ok := allRules["*"]
+		if ok && slices.Contains(starRules, "get") || slices.Contains(starRules, "*") {
+			return parsed, nil
+		}
+
+		for mcName := range allRules {
+			// Should have "GET" auth for all managedCluster
+			if getAccessByClusterName(allRules, mcName) {
+				parsed.Filters["clusters.name"] = append(parsed.Filters["clusters.name"], mcName)
+			}
+		}
+
+		if len(parsed.Filters["clusters.name"]) == 0 {
+			return parsed, ErrNoAccess
+		}
+	}
+
+	if len(unAuthorizedClusters) > 0 {
+		return parsed, fmt.Errorf("%w: the following cluster filters are not authorized: %s",
+			ErrNotAuthorized, strings.Join(unAuthorizedClusters, ", "))
```

Comment:
I think we should have a new error called `ErrForbidden` to match the HTTP status codes we return. It's getting a bit confusing to use this for both cases.